param.go

```diff
@@ -326,3 +296,126 @@ func (po paramObject) Build(c *Container) (reflect.Value, error) {
 	}
 	return dest, nil
 }
+
+// paramObjectField is a single field of a dig.In struct.
+type paramObjectField struct {
+	// Name of the field in the struct.
+	FieldName string
+
+	// Index of this field in the target struct.
+	//
+	// We need to track this separately because not all fields of the
+	// struct map to params.
+	FieldIndex int
+
+	// The dependency requested by this field.
+	Param param
+}
+
+func newParamObjectField(idx int, f reflect.StructField) (paramObjectField, error) {
+	pof := paramObjectField{
+		FieldName:  f.Name,
+		FieldIndex: idx,
+	}
+
+	var p param
+	switch {
+	case f.PkgPath != "":
+		return pof, fmt.Errorf(
+			"unexported fields not allowed in dig.In, did you mean to export %q (%v)?",
+			f.Name, f.Type)
+
+	case f.Tag.Get(_groupTag) != "":
+		var err error
+		p, err = newParamGroupedSlice(f)
+		if err != nil {
+			return pof, err
+		}
+
+	default:
+		var err error
+		p, err = newParam(f.Type)
+		if err != nil {
+			return pof, err
+		}
+	}
+
+	if ps, ok := p.(paramSingle); ok {
+		ps.Name = f.Tag.Get(_nameTag)
+
+		var err error
+		ps.Optional, err = isFieldOptional(f)
+		if err != nil {
+			return pof, err
+		}
+
+		p = ps
+	}
+
+	pof.Param = p
+	return pof, nil
+}
+
+// paramGroupedSlice is a param which produces a slice of values with the same
+// group name.
+type paramGroupedSlice struct {
+	// Name of the group as specified in the `group:".."` tag.
+	Group string
+
+	// Type of the slice.
+	Type reflect.Type
+}
+
+// newParamGroupedSlice builds a paramGroupedSlice from the provided type with
+// the given name.
+//
+// The type MUST be a slice type.
+func newParamGroupedSlice(f reflect.StructField) (paramGroupedSlice, error) {
+	pg := paramGroupedSlice{Group: f.Tag.Get(_groupTag), Type: f.Type}
+
+	name := f.Tag.Get(_nameTag)
+	optional, _ := isFieldOptional(f)
+	switch {
+	case f.Type.Kind() != reflect.Slice:
+		return pg, fmt.Errorf("value groups may be consumed as slices only: "+
+			"field %q (%v) is not a slice", f.Name, f.Type)
+	case name != "":
+		return pg, fmt.Errorf(
+			"cannot use named values with value groups: name:%q provided with group:%q", name, pg.Group)
+
+	case optional:
+		return pg, errors.New("value groups cannot be optional")
+	}
+
+	return pg, nil
+}
+
+func (pt paramGroupedSlice) Build(c *Container) (reflect.Value, error) {
+	k := key{group: pt.Group, t: pt.Type.Elem()}
+
+	for _, n := range c.providers[k] {
+		if err := n.Call(c); err != nil {
+			// TODO: better message
+			return _noValue, errWrapf(err, "failed to build %v", k)
```

Comment:
Can we include the group name in the error message? (Or is it included further up the error stack?)

Reply:
The message right now will be,

    failed to build [$type]:$groupName: ...

Reply:
Since we aren't allowed to use optional, what's the behavior for a request value group with no provided types? I'm assuming that this would be an error and that this is the point of failure. We should consider adding a note to the package-level documentation about this.

Reply:
If you request `group:"foo"` and nothing provided values to it, you'll get an empty slice.

(Is that what you're asking?)